beacon-chain/sync/rpc_blob_sidecars_by_range.go

```diff
@@ -0,0 +1,179 @@
+		return err
+	}
+	rp, err := validateBlobsByRange(r, s.cfg.chain.CurrentSlot())
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	// Ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	batcher, err := newBlockRangeBatcher(rp, s.cfg.beaconDB, s.rateLimiter, s.cfg.chain.IsCanonical, ticker)
+
+	var batch blockBatch
+	var totalWrites uint64
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
+		batchStart := time.Now()
+		writes, err := s.streamBlobBatch(ctx, batch, stream)
+		rpcBlobsByRangeResponseLatency.Observe(float64(time.Since(batchStart).Milliseconds()))
+		if err != nil {
+			return err
+		}
+		totalWrites += writes
+		// once we have written MAX_REQUEST_BLOB_SIDECARS, we're done serving the request
+		if totalWrites >= params.BeaconNetworkConfig().MaxRequestBlobsSidecars {
+			break
+		}
+	}
+	if err := batch.error(); err != nil {
+		log.WithError(err).Debug("error in BlocksByRange batch")
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	closeStream(stream, log)
+	return nil
+}
+
+func blobsByRangeMinStartSlot(current primitives.Slot) (primitives.Slot, error) {
+	minReqEpochs := params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
+	currEpoch := slots.ToEpoch(current)
+	minStart := params.BeaconConfig().DenebForkEpoch
+	if currEpoch > minReqEpochs && currEpoch-minReqEpochs > minStart {
+		minStart = currEpoch - minReqEpochs
+	}
+	return slots.EpochStart(minStart)
+}
+
+func blobBatchLimit() uint64 {
+	return uint64(flags.Get().BlockBatchLimit / fieldparams.MaxBlobsPerBlock)
+}
+
+func validateBlobsByRange(r *pb.BlobSidecarsByRangeRequest, current primitives.Slot) (rangeParams, error) {
+	if r.Count == 0 {
+		return rangeParams{}, errors.Wrap(p2ptypes.ErrInvalidRequest, "invalid request Count parameter")
+	}
+	rp := rangeParams{
+		start: r.StartSlot,
+		size:  r.Count,
+	}
+	// Peers may overshoot the current slot when in initial sync, so we don't want to penalize them by treating the
+	// request as an error. So instead we return a set of params that acts as a noop.
+	if rp.start > current {
+		return rangeParams{start: current, end: current, size: 0}, nil
+	}
+
+	var err error
+	rp.end, err = rp.start.SafeAdd((rp.size - 1))
+	if err != nil {
+		return rangeParams{}, errors.Wrap(p2ptypes.ErrInvalidRequest, "overflow start + count -1")
+	}
+
+	maxRequest := params.BeaconNetworkConfig().MaxRequestBlocksDeneb
+	// Allow some wiggle room, up to double the MaxRequestBlocks past the current slot,
+	// to give nodes syncing close to the head of the chain some margin for error.
+	maxStart, err := current.SafeAdd(maxRequest * 2)
```

Comment:
does this limit apply for blobs too ? I noticed that blobs aren't to be served until a minimum threshold is reached, unlike blocks.

Reply:
Updating this discussion from our chat elsewhere - the reasoning here is that we should use the same start slot validation logic as `BeaconBlocksByRange` so that we don't penalize clients who try to sync blobs in parallel with blocks (and potentially overshoot current slot for both). Ultimately we won't serve any results when the start slot is > current, but we also won't treat a request like this as an error and penalize clients for aggressively seeking past this point, within the `maxRequest * 2` margin of error.